### ISSUE TITLE
Cannot load my engine - failed to start node

### ISSUE THREAD

**Issue by mzeidhassan** [ModernMT/MMT#329]
Hello MMT,

Not sure what I am having some issues with MMT, especially after porting my old engine. 

Here is what I see in the terminal after running ./mmt start

> Starting MMT engine 'default'... OK
> Loading models... FAIL
> ERROR Unexpected exception:
> 	Exception('failed to start node, check log file for more details: /home/mlearner/ModernMT/runtime/default/logs/node.log',)

and here is the node.log output. 

By the way, I have done clean build already using these commands:

`git checkout -- .
git clean -fd .

rm -rf build/*
git pull origin master

cd vendor
./compile
cd ../src
mvn clean install`

I also noticed something else:

According to [this issue](https://github.com/ModernMT/MMT/issues/322#issuecomment-350961165) , the command was:

`extras/porting/convert_from_v2.0.2_to_v2.0.3 --engine-path engines/default`

But now I see a different script  `convert_from_v2.0.2_to_v2.1`in 'ModernMT/extras/porting'

Any idea?

I hope there is a quick fix. 

Thanks,
Mohamed

> 2017-12-16 12:55:23,048 INFO  eu.modernmt.cluster.ClusterNode [main] Node is joining the cluster
> 2017-12-16 12:55:23,209 INFO  com.hazelcast.instance.AddressPicker [main] [LOCAL] [eu.modernmt] [3.9.1] Interfaces is disabled, trying to pick one address from TCP-IP config addresses: []
> 2017-12-16 12:55:23,209 INFO  com.hazelcast.instance.AddressPicker [main] [LOCAL] [eu.modernmt] [3.9.1] Prefer IPv4 stack is true.
> 2017-12-16 12:55:23,209 WARN  com.hazelcast.instance.AddressPicker [main] [LOCAL] [eu.modernmt] [3.9.1] Could not find a matching address to start with! Picking one of non-loopback addresses.
> 2017-12-16 12:55:23,213 INFO  com.hazelcast.instance.AddressPicker [main] [LOCAL] [eu.modernmt] [3.9.1] Picked [172.17.0.1]:5016, using socket ServerSocket[addr=/0:0:0:0:0:0:0:0,localport=5016], bind any local is true
> 2017-12-16 12:55:23,213 INFO  com.hazelcast.instance.AddressPicker [main] [LOCAL] [eu.modernmt] [3.9.1] Using public address: [127.0.1.1]:5016
> 2017-12-16 12:55:23,218 INFO  com.hazelcast.system [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Hazelcast 3.9.1 (20171130 - feca534) starting at [127.0.1.1]:5016
> 2017-12-16 12:55:23,218 INFO  com.hazelcast.system [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
> 2017-12-16 12:55:23,218 INFO  com.hazelcast.system [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Configured Hazelcast Serialization version: 1
> 2017-12-16 12:55:23,284 INFO  com.hazelcast.spi.impl.operationservice.impl.BackpressureRegulator [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Backpressure is disabled
> 2017-12-16 12:55:23,472 INFO  com.hazelcast.instance.Node [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Creating TcpIpJoiner
> 2017-12-16 12:55:23,532 INFO  com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Starting 8 partition threads and 5 generic threads (1 dedicated for priority tasks)
> 2017-12-16 12:55:23,534 INFO  com.hazelcast.internal.diagnostics.Diagnostics [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
> 2017-12-16 12:55:23,536 INFO  com.hazelcast.core.LifecycleService [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] [127.0.1.1]:5016 is STARTING
> 2017-12-16 12:55:23,543 INFO  com.hazelcast.system [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] Cluster version set to 3.9
> 2017-12-16 12:55:23,543 INFO  com.hazelcast.internal.cluster.ClusterService [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] 
> 
> Members {size:1, ver:1} [
> 	Member [127.0.1.1]:5016 - 02b7ed04-9a06-4508-9fad-8e24088d7eb6 this
> ]
> 
> 2017-12-16 12:55:23,553 INFO  com.hazelcast.core.LifecycleService [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] [127.0.1.1]:5016 is STARTED
> 2017-12-16 12:55:23,554 INFO  eu.modernmt.cluster.ClusterNode [main] Node joined the cluster in 0.364s
> 2017-12-16 12:55:23,554 WARN  eu.modernmt.cluster.ClusterNode [main] Model syncing not supported in this version!
> 2017-12-16 12:55:23,555 INFO  eu.modernmt.cluster.ClusterNode [main] Model loading started
> 2017-12-16 12:55:23,841 INFO  eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (nmmt.NMTDecoder) Loading "en__ar" model from checkpoint... START
> 2017-12-16 12:55:28,339 INFO  eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (nmmt.NMTDecoder) Loading "en__ar" model from checkpoint END 4.50s
> 2017-12-16 12:55:28,339 ERROR eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (root) __init__() got an unexpected keyword argument 'dim'
> Traceback (most recent call last):
>   File "main_loop.py", line 171, in run_main
>     decoder = NMTDecoder(args.model, gpu_id=args.gpu, random_seed=3435)
>   File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTDecoder.py", line 66, in __init__
>     self._engines[key].running_state = NMTEngine.HOT
>   File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTEngine.py", line 384, in running_state
>     self.__load()
>   File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTEngine.py", line 162, in __load
>     decoder = Models.Decoder(self.metadata, self.trg_dict)
>   File "/home/mlearner/ModernMT/build/lib/pynmt/onmt/Models.py", line 119, in __init__
>     self.attn = onmt.modules.GlobalAttention(opt.rnn_size)
>   File "/home/mlearner/ModernMT/build/lib/pynmt/onmt/modules/GlobalAttention.py", line 31, in __init__
>     self.sm = nn.Softmax(dim=1)
> TypeError: __init__() got an unexpected keyword argument 'dim'
> 2017-12-16 12:55:28,419 FATAL eu.modernmt.facade.ModernMT [main] Unexpected exception thrown by thread [main]
> java.lang.NullPointerException
> 	at java.io.StringReader.<init>(StringReader.java:50)
> 	at com.google.gson.JsonParser.parse(JsonParser.java:45)
> 	at eu.modernmt.decoder.neural.execution.NativeProcess.deserialize(NativeProcess.java:225)
> 	at eu.modernmt.decoder.neural.execution.NativeProcess.<init>(NativeProcess.java:134)
> 	at eu.modernmt.decoder.neural.execution.NativeProcess$Builder.start(NativeProcess.java:101)
> 	at eu.modernmt.decoder.neural.execution.NativeProcess$Builder.startOnGPU(NativeProcess.java:75)
> 	at eu.modernmt.decoder.neural.execution.StartNativeProcessGpuTask.call(StartNativeProcessGpuTask.java:34)
> 	at eu.modernmt.decoder.neural.execution.SingletonExecutionQueue.forGPU(SingletonExecutionQueue.java:47)
> 	at eu.modernmt.decoder.neural.execution.ExecutionQueue.newGPUInstance(ExecutionQueue.java:29)
> 	at eu.modernmt.decoder.neural.NeuralDecoder.<init>(NeuralDecoder.java:65)
> 	at eu.modernmt.engine.impl.NeuralEngine.<init>(NeuralEngine.java:32)
> 	at eu.modernmt.engine.Engine.load(Engine.java:70)
> 	at eu.modernmt.cluster.ClusterNode.start(ClusterNode.java:314)
> 	at eu.modernmt.facade.ModernMT.start(ModernMT.java:49)
> 	at eu.modernmt.cli.ClusterNodeMain.main(ClusterNodeMain.java:168)

**Comment by AndRossi** [comment]
Hi @mzeidhassan, 

I believe I know what is happening there, and there should be an immediate solution :)  

Starting from the version 2.1 that we released two days ago, MMT uses CUDA 9 and PyTorch 0.3. So, when you update MMT to version 2.1, in addition to porting your old engine with the porting script (you have already done this) you should also update these softwares.

I have written all the information you need in the end of [this section]
(https://github.com/ModernMT/MMT/wiki/Upgrading-MMT#porting-20x-engines) 
Of course if you have any doubt, don't hesitate to ask :)

Thank you for your patience!

Andrea

**Comment by mzeidhassan** [comment]
@AndRossi Thanks Andrea for getting back to me.

I have a couple of old engines running for testing purposes and I don't plant to port them now. If I update to Cuda 9 and Pytorch 0.3, do you think this will break these engines? 

Thanks,
Mohamed

**Comment by AndRossi** [comment]
Hi @mzeidhassan, 

As long as you do not update the version of the MMT software that your old engines are using, they should not be affected by updating CUDA drivers and PyTorch :)

Andrea

**Comment by mzeidhassan** [comment]
OK, this is great to know! Thanks @AndRossi !

**Comment by mzeidhassan** [comment]
Hi @AndRossi,

Unfortunately, updating pytorch and cuda drivers didn't fix the issue. I still get this issue:

> sudo extras/porting/convert_from_v2.0.2_to_v2.1 --engine-path engines/default
Creating decoder backup... Done
Updating decoder configuration... An error occurred. Reverting the backup data into the decoder... Done
Traceback (most recent call last):
  File "extras/porting/convert_from_v2.0.2_to_v2.1", line 107, in <module>
    main()
  File "extras/porting/convert_from_v2.0.2_to_v2.1", line 84, in main
    raise e
AttributeError: 'NoneType' object has no attribute 'strip'

For some reason, the porting script doesn't work.

Any idea?
Thanks,
Mohamed

**Comment by mzeidhassan** [comment]
Hi @AndRossi ,

I think I know why I am having this issue. 

As you may remember [](https://github.com/ModernMT/MMT/issues/322), I have ported my old engine '2.0.2 ' to '2.0.3' using the old script `'convert_from_v2.0.2_to_v2.0.3'`. So, I think the new porting script doesn't work this way and it needs a 2.0.2 engine to work. I had a backup of my old engine, I replaced the 2.0.3 engine with it, run your script 'convert_from_v2.0.2_to_v2.1' and it worked just fine and I got this message: 

Congratulations! Your 2.0.2 engine was successfully ported to version 2.1

So, is there a way to port my 2.0.3 engine to 2.1? Can you create a script for this? I don't want to lose the data in that engine for now. I hope this is doable.

Thanks,
Mohamed

**Comment by mzeidhassan** [comment]
Unfortunately, even with the successful porting of the model, it still doesn't load. Here is the node.log text:

> 2017-12-20 00:31:23,424 INFO  com.hazelcast.core.LifecycleService [main] [127.0.1.1]:5016 [eu.modernmt] [3.9.1] [127.0.1.1]:5016 is STARTED
2017-12-20 00:31:23,426 INFO  eu.modernmt.cluster.ClusterNode [main] Node joined the cluster in 0.565s
2017-12-20 00:31:23,429 WARN  eu.modernmt.cluster.ClusterNode [main] Model syncing not supported in this version!
2017-12-20 00:31:23,430 INFO  eu.modernmt.cluster.ClusterNode [main] Model loading started
2017-12-20 00:31:23,855 INFO  eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (nmmt.NMTDecoder) Loading "en__ar" model from checkpoint... START
2017-12-20 00:31:28,366 INFO  eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (nmmt.NMTDecoder) Loading "en__ar" model from checkpoint END 4.51s
2017-12-20 00:31:28,366 ERROR eu.modernmt.decoder.neural.execution.NativeProcess [Thread-4] (root) __init__() got an unexpected keyword argument 'dim'
Traceback (most recent call last):
  File "main_loop.py", line 171, in run_main
    decoder = NMTDecoder(args.model, gpu_id=args.gpu, random_seed=3435)
  File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTDecoder.py", line 66, in __init__
    self._engines[key].running_state = NMTEngine.HOT
  File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTEngine.py", line 384, in running_state
    self.__load()
  File "/home/mlearner/ModernMT/build/lib/pynmt/nmmt/NMTEngine.py", line 162, in __load
    decoder = Models.Decoder(self.metadata, self.trg_dict)
  File "/home/mlearner/ModernMT/build/lib/pynmt/onmt/Models.py", line 119, in __init__
    self.attn = onmt.modules.GlobalAttention(opt.rnn_size)
  File "/home/mlearner/ModernMT/build/lib/pynmt/onmt/modules/GlobalAttention.py", line 31, in __init__
    self.sm = nn.Softmax(dim=1)
TypeError: __init__() got an unexpected keyword argument 'dim'
2017-12-20 00:31:28,446 FATAL eu.modernmt.facade.ModernMT [main] Unexpected exception thrown by thread [main]
java.lang.NullPointerException
	at java.io.StringReader.<init>(StringReader.java:50)
	at com.google.gson.JsonParser.parse(JsonParser.java:45)
	at eu.modernmt.decoder.neural.execution.NativeProcess.deserialize(NativeProcess.java:225)
	at eu.modernmt.decoder.neural.execution.NativeProcess.<init>(NativeProcess.java:134)
	at eu.modernmt.decoder.neural.execution.NativeProcess$Builder.start(NativeProcess.java:101)
	at eu.modernmt.decoder.neural.execution.NativeProcess$Builder.startOnGPU(NativeProcess.java:75)
	at eu.modernmt.decoder.neural.execution.StartNativeProcessGpuTask.call(StartNativeProcessGpuTask.java:34)
	at eu.modernmt.decoder.neural.execution.SingletonExecutionQueue.forGPU(SingletonExecutionQueue.java:47)
	at eu.modernmt.decoder.neural.execution.ExecutionQueue.newGPUInstance(ExecutionQueue.java:29)
	at eu.modernmt.decoder.neural.NeuralDecoder.<init>(NeuralDecoder.java:65)
	at eu.modernmt.engine.impl.NeuralEngine.<init>(NeuralEngine.java:32)
	at eu.modernmt.engine.Engine.load(Engine.java:70)
	at eu.modernmt.cluster.ClusterNode.start(ClusterNode.java:314)
	at eu.modernmt.facade.ModernMT.start(ModernMT.java:49)
	at eu.modernmt.cli.ClusterNodeMain.main(ClusterNodeMain.java:168)

What could be the issue?

Thanks,
Mohamed

**Comment by davidecaroselli** [comment]
Hi @mzeidhassan !

If you want to use new MMT v2.1 with its new engines (or ported one of course), you have to install PyTorch 0.3.0 otherwise you will get the error you're seeing: `got an unexpected keyword argument 'dim'`

Cheers,
Davide

**Comment by davidecaroselli** [comment]
Sorry, just read this:
>So, is there a way to port my 2.0.3 engine to 2.1? Can you create a script for this? I don't want to lose the data in that engine for now. I hope this is doable.

The "old" conversion script is exactly the same of the "new" one, we just changed the name :) So the engine you ported with previous script must be compatible with the MMT v2.1!

**Comment by mzeidhassan** [comment]
Hi @davidecaroselli ,

This is what I see when I type this in Python terminal.

>>> import torch
>>> print(torch.__version__)
0.3.0.post4

So, it seems I am having the correct version.

As for Cuda, this is what I see after following your instruction at https://github.com/ModernMT/MMT/blob/master/INSTALL.md#cuda-drivers-required-for-neural-engine

sudo apt-get install cuda
Reading package lists... Done
Building dependency tree       
Reading state information... Done
cuda is already the newest version (9.1.85-1).

So, I am not sure what it doesn't load, even after a successful porting.

**Comment by davidecaroselli** [comment]
Hmmm let's see: can you confirm that you have installed PyTorch 0.3 for python 2.x and not 3.x?
Let's do this test: at the very end of the file `cli/mmt/neural.py` add this line:
```python
print "\n\n PyTorch version is " + torch.__version__ + "\n\n"
```

Then try to start the engine ad see which version of PyTorch is MMT actually using.

**Comment by mzeidhassan** [comment]
Here is what I see

PyTorch version is 0.2.0_3

By the way, I am using Anaconda with Python 3.6 and I installed Cuda using this command from pytorch page:

conda install pytorch torchvision cuda90 -c pytorch

**Comment by mzeidhassan** [comment]
If I use this:
pip install http://download.pytorch.org/whl/cu90/torch-0.3.0.post4-cp27-cp27mu-linux_x86_64.whl 

I get this error:

`torch-0.3.0.post4-cp27-cp27mu-linux_x86_64.whl is not a supported wheel on this platform.`

**Comment by davidecaroselli** [comment]
Ok then, that's definitely the problem: MMT is using a python version that does not have PyTorch 0.3.0. I'm not very familiar with Anaconda, but I know users who work with MMT under Anaconda.

You need to either install PyTorch 0.3.0 in your regular python environment or make MMT run under Anaconda (is there some sort of env activation?). Unfortunately I cannot help you with conda.

I Googled the error `torch-0.3.0 is not a supported wheel on this platform` and it seems related to a misconfiguration of pip with Anaconda or something like that, but it's very difficult for me to understand what's the configuration problem on your machine.

If there's something I can help you with, please let me know!

**Comment by mzeidhassan** [comment]
@davidecaroselli I am using the root env of Anaconda and it's the only Python version installed. So, you think it's better to uninstall anaconda completely and install regular Python?

**Comment by mzeidhassan** [comment]
@davidecaroselli one more question:

So, MMT can work with Python 3.6, Pytorch 0.3.0? Or do I need Python 2.7 to make it work? 

Thanks,
Mohamed

**Comment by davidecaroselli** [comment]
>So, MMT can work with Python 3.6, Pytorch 0.3.0? Or do I need Python 2.7 to make it work?

MMT only works with Python 2.7. With the Anaconda environment, do you have PyTorch 0.3.0 installed also on Python 2.7? If not, try installing it and the run again MMT. Maybe Anaconda is fine, the problem is just that PyTorch 0.3.0 is not available for the right version of python.

**Comment by mzeidhassan** [comment]
Thanks @davidecaroselli ! 

OK, I have now installed the Pytorch 0.3.0 for Python 2.7 using this command:

`sudo python2.7 /usr/bin/pip install http://download.pytorch.org/whl/cu90/torch-0.3.0.post4-cp27-cp27mu-linux_x86_64.whl`

I had to specify the version to be able to install it to Python 2.7. 

Now everything is working just fine and when I run mmt, I see now the correct version:

"PyTorch version is 0.3.0.post4"

I hope this will help someone else who may struggle with this issue. This was a headache indeed, but I am happy that it all ended well.

Thanks again Davide for your support.
Mohamed
